src/modules/rlm_mschap/rlm_mschap.c

```diff
@@ -560,82 +580,77 @@ static ssize_t mschap_xlat(UNUSED TALLOC_CTX *ctx, char **out, size_t outlen,
 			} else {
 				p = user_name->vp_strvalue; /* use the whole User-Name */
 			}
-			strlcpy(*out, p, outlen);
+			fr_value_box_strdup(ctx, vb, NULL, p, true);
 		}
 
-		return strlen(*out);
+		fr_dcursor_append(out, vb);
+		return XLAT_ACTION_DONE;
 
 	/*
 	 * Return the NT-Hash of the passed string
 	 */
-	} else if (strncasecmp(fmt, "NT-Hash ", 8) == 0) {
-		char const *p;
-
-		p = fmt + 8;	/* 7 is the length of 'NT-Hash' */
-		if ((*p == '\0') || (outlen <= 32))
-			return 0;
+	} else if (strncasecmp(arg->vb_strvalue, "NT-Hash", 7) == 0) {
+		char	*p;
 
-		fr_skip_whitespace(p);
+		arg = fr_dlist_next(in, arg);
+		if ((!arg) || (arg->length == 0))
+			return XLAT_ACTION_FAIL;
 
-		if (mschap_nt_password_hash(buffer, p) < 0) {
+		if (mschap_nt_password_hash(buffer, arg->vb_strvalue) < 0) {
 			REDEBUG("Failed generating NT-Password");
 			*buffer = '\0';
-			return -1;
+			return XLAT_ACTION_FAIL;
 		}
 
-		fr_bin2hex(&FR_SBUFF_OUT(*out, (NT_DIGEST_LENGTH * 2) + 1), &FR_DBUFF_TMP(buffer, NT_DIGEST_LENGTH), SIZE_MAX);
-		(*out)[32] = '\0';
-		RDEBUG2("NT-Hash of \"known-good\" password: %s", *out);
-		return 32;
+		vb = fr_value_box_alloc_null(ctx);
+		fr_value_box_bstr_alloc(ctx, &p, vb, NULL, (NT_DIGEST_LENGTH *2), false);
+		fr_bin2hex(&FR_SBUFF_OUT(p, (NT_DIGEST_LENGTH * 2) + 1), &FR_DBUFF_TMP(buffer, NT_DIGEST_LENGTH), SIZE_MAX);
```

Comment:
This, along with any other sub-functions which emit binary data, should return an octets type box, instead of a hex string.